src/server/debug/server/server.go

```diff
@@ -208,73 +143,418 @@ func (s *debugServer) appLogs(ctx context.Context, tw *tar.Writer) (retErr error
+							}
+							c, err := workerserver.NewClient(pod.Ip)
+							if err != nil {
+								return errors.Wrapf(err, "worker client for %q at %q", pod.Name, pod.Ip)
+							}
+							prof, err := c.Profile(ctx, req)
+							if err != nil {
+								return errors.Wrapf(err, "collect profile %q for %q at %q", profile, pod.Name, pod.Ip)
+							}
+							r := grpcutil.NewStreamingBytesReader(prof, nil)
+							defer func() {
+								if err := r.Close(); err != nil {
+									errors.JoinInto(&retErr, errors.Wrapf(err, "close profile writer for pod %q; profile %q", pod.Name, profile))
+								}
+							}()
+							_, err = io.Copy(w, r)
+							return err
+						}); err != nil {
+							errors.JoinInto(&errs, err)
+						}
+					}
+				}
+			}
+			rp(ctx)
+		}
+		log.Error(ctx, "profile errors", zap.Error(errs))
+		return errs
+
+	}
+}
+
+func makeBinariesTask(server debug.Debug_DumpV2Server, apps []*debug.App) taskFunc {
+	return func(ctx context.Context, dfs DumpFS) error {
+		var errs error
+		rp := recordProgress(server, "binaries", len(apps))
+		for _, app := range apps {
+			for _, pod := range app.Pods {
+				if pod.Ip == "" {
+					log.Info(ctx, "skipping debug profile for pod with empty IP", zap.String("pod", pod.Name))
+					continue
+				}
+				if err := dfs.Write(filepath.Join(appDir(app), "pods", pod.Name, "binary"), func(w io.Writer) (retErr error) {
+					if app.Name == "pachd" {
+						f, err := os.Open(os.Args[0])
+						if err != nil {
+							return errors.Wrap(err, "open file")
+						}
+						defer func() {
+							if err := f.Close(); err != nil {
+								errors.JoinInto(&retErr, err)
+							}
+						}()
+						_, err = io.Copy(w, f)
+						return err
+					}
+					c, err := workerserver.NewClient(pod.Ip)
+					if err != nil {
+						return errors.Wrapf(err, "create worker client for IP %q", pod.Ip)
+					}
+					binaryC, err := c.Binary(ctx, &debug.BinaryRequest{})
+					if err != nil {
+						return errors.Wrap(err, "binary client")
+					}
+					r := grpcutil.NewStreamingBytesReader(binaryC, nil)
+					defer func() {
+						if err := r.Close(); err != nil {
+							errors.JoinInto(&retErr, errors.Wrapf(err, "close binary writer for pod %q", pod.Name))
+						}
+					}()
+					_, err = io.Copy(w, r)
+					return err
+				}); err != nil {
+					errors.JoinInto(&errs, errors.Wrapf(err, "close binary writer for pod %q", pod.Name))
+				}
+			}
+			rp(ctx)
+		}
+		return errs
+	}
+}
+
+func appDir(app *debug.App) string {
+	if app.Pipeline != nil {
+		return filepath.Join("pipelines", app.Pipeline.Project, app.Pipeline.Name)
+	}
+	return app.Name
+}
+
+type dumpContentServer struct {
+	server debug.Debug_DumpV2Server
+}
+
+func (dcs *dumpContentServer) Send(bytesValue *types.BytesValue) error {
+	return dcs.server.Send(&debug.DumpChunk{Chunk: &debug.DumpChunk_Content{Content: &debug.DumpContent{Content: bytesValue.Value}}})
+}
+
+type incProgressFunc func(context.Context)
+
+// TODO should this be sending in a goro?
```

Comment:
Only one goroutine can call Send at a time.  If we parallelize dumping this will need additional synchronization.

Reply:
so dumping is currently being parallelized. It doesn't just block on a channel?

Reply:
go doc grpc.ServerStream says:

>         // It is safe to have a goroutine calling SendMsg and another goroutine
>        // calling RecvMsg on the same stream at the same time, but it is not safe
>        // to call SendMsg on the same stream in different goroutines.